### PR TITLE
Add sequencer staking rewards tests

### PR DESCRIPTION
**Pull request by AlexChetverov**

### CODE REVIEW

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+  }
+  await setupApi();
+  setupUsers();
+});
+
+beforeEach(async () => {
+  await setupApi();
+  [testUser] = setupUsers();
+  testUser.addAsset(GASP_ASSET_ID);
+  api = getApi();
```
Comment (goncer) [inline test/sequential/sequencerStaking.rewards.test.ts:63]: I think this is not used anymore

File: utils/rollDown/Rolldown.ts [hunk 2]
```diff
@@ -234,10 +234,9 @@ export class Rolldown {
 
   static getRequestIdFromEvents(
     events: MangataGenericEvent[],
-    module = "rolldown",
     method = "L1ReadStored",
   ) {
-    const event = getEventResultFromMangataTx(events, [module, method]);
+    const event = getEventResultFromMangataTx(events, [method]);
```
Comment (goncer) [inline utils/rollDown/Rolldown.ts:239]: why this change?

**Review by goncer — CHANGES_REQUESTED** [review]

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+import { AssetWallet, User } from "../../utils/User";
+import { GASP_ASSET_ID } from "../../utils/Constants";
+import { ApiPromise } from "@polkadot/api";
+
+let testUser: User;
+let api: ApiPromise;
+const chainEth = "Ethereum";
+const chainArb = "Arbitrum";
+
+async function getUpdate(txIndex: number, userAddress: string, chain: string) {
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:40]: Done

File: test/sequential/sequencerStaking.rewards.test.ts (hunk 1 above)
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:63]: Changed

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+it("Sequencer budget is set when initializing issuance config", async () => {
+  const events = await Sudo.batchAsSudoFinalized(
+    Assets.FinalizeTge(),
+    Assets.initIssuance(),
+  );
+  const filteredEvent = await filterAndStringifyFirstEvent(
+    events,
+    "IssuanceConfigInitialized",
+  );
+  expect(filteredEvent[0].sequencersSplit).not.toBeEmpty();
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:86]: We have a problem with it, because we recieved the value as percent. Percentage value used

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+});
+
+it("Sequencers get paid on every session BUT only when they submit valid updates ( Succeeded extrinsics )", async () => {
+  await Sudo.batchAsSudoFinalized(Assets.FinalizeTge(), Assets.initIssuance());
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:90]: I understand, but in this case the tests will be codependent, which is not very good. I thought about running these functions in BeforeAll, creating a variable for the event there, and then using it in the first test, but I didn't know how you would rate this idea.
let initEvent: MangataGenericEvent[];
beforeAll(async () => {
      initEvents = await Sudo.batchAsSudoFinalized(
    Assets.FinalizeTge(),
    Assets.initIssuance(),
  );
  ...
})
...
it("Sequencer budget is set when initializing issuance config", async () => {
      const filteredEvent = await filterAndStringifyFirstEvent(
    initEvents,
    "IssuanceConfigInitialized",
  );
...
})

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  const { reqId } = await createAnUpdate(testUser, chainEth);
+  const rewardsSessionNumber = await getSessionIndex();
+  const registrationBlock = reqId + 1;
+  await waitBlockNumber(registrationBlock.toString(), disputePeriodLength * 2);
+  await waitNewStakingRound();
+  //We receive rewards data only two rounds after the update
+  const sessionIndex = await getSessionIndex();
+  if (sessionIndex <= rewardsSessionNumber + 1) {
+    await waitNewStakingRound();
+  }
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:101]: Done

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+it("When a sequencer brings an update It will get some points", async () => {
+  await Sudo.batchAsSudoFinalized(Assets.FinalizeTge(), Assets.initIssuance());
+  await createAnUpdate(testUser, chainEth);
+  const rewardsSessionNumber = await getSessionIndex();
+  const pointsValue = await SequencerStaking.points(rewardsSessionNumber);
+  const userAwardedPts = await SequencerStaking.awardedPts(
+    rewardsSessionNumber,
+    testUser.keyRingPair.address,
+  );
+  expect(userAwardedPts).bnEqual(pointsValue);
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:140]: Done

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+  const txIndexEth = await Rolldown.lastProcessedRequestOnL2(chainEth);
+  const txIndexArb = await Rolldown.lastProcessedRequestOnL2(chainArb);
+  await Rolldown.waitForReadRights(testUser3.keyRingPair.address, 50, chainArb);
+  await waitNewStakingRound();
+  const selectedSequencerEth = (
+    await getApi().query.sequencerStaking.selectedSequencer()
+  ).toHuman().Ethereum;
+  let ethUser1: User;
+  let ethUser2: User;
+  if (selectedSequencerEth!.toString() === testUser.keyRingPair.address) {
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:173]: The problem is that we need to make the first deposit in the first half of the session, but we can't be sure that the first Ethereum sequencer will be the selected sequencer in the first half of the session. So an option was added to determine which sequencer is the first selected sequencer in the session. Added a comment

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+      await getUpdate(txIndexEth, ethUser1.keyRingPair.address, chainEth),
+    ),
+    Sudo.sudoAsWithAddressString(
+      testUser3.keyRingPair.address,
+      await getUpdate(txIndexArb, testUser3.keyRingPair.address, chainArb),
+    ),
+  );
+  const waitingBlockNumber =
+    (await Rolldown.getRequestIdFromEvents(updateEvents)) + 1;
+  await waitBlockNumber(waitingBlockNumber.toString(), 50);
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:194]: I found a similar function and added it

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+  const rewardsSessionNumber = await getSessionIndex();
+  await Sudo.batchAsSudoFinalized(
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:196]: We've discussed a scenario in which we perform an operation in the Ethereum network with different selected sequencers in the Arbitrum network. I assume that since the time of the dispute is equal to half the time of the session, the first function is performed at the very beginning and the second after the period of the dispute, all this happens in one session. We also check this by counting the rewards at the end of the test.

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+    testUser3.keyRingPair.address,
+  );
+  expect(user1AwardedPts).bnEqual(pointsValue.divn(3));
+  expect(user2AwardedPts).bnEqual(BN_ZERO);
+  expect(user3AwardedPts).bnEqual(pointsValue.divn(3).muln(2));
+  let sessionIndex = await getSessionIndex();
+  while (sessionIndex <= rewardsSessionNumber + 1) {
+    await waitNewStakingRound();
+    sessionIndex = await getSessionIndex();
+  }
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:222]: Done

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+  await leaveSequencing(testUser.keyRingPair.address);
+  const sequencersList = await SequencerStaking.activeSequencers();
+  expect(sequencersList.toHuman().Ethereum).not.toContain(
+    testUser.keyRingPair.address,
+  );
+  let sessionIndex = await getSessionIndex();
+  while (sessionIndex <= rewardsSessionNumber + 1) {
+    await waitNewStakingRound();
+    sessionIndex = await getSessionIndex();
+  }
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:285]: When we divide the reward into three parts it may turn out that the reward is equal, then when dividing into 3 parts and 2 users we will get numbers that differ by one. I am not completely sure how this works in the system, but I received a reward for one and another sequencer, say, 630001 and 315000, while I wanted to leave the most correct check that these are not random numbers, namely 2/3 and 1/3 of the total reward

File: test/sequential/sequencerStaking.rewards.test.ts [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+  const diff1 = ethUser1.getWalletDifferences()[0].diff.free;
+  const diff2 = ethUser2.getWalletDifferences();
+  const diff3 = testUser3.getWalletDifferences()[0].diff.free;
+  expect(diff3).bnGt(diff1);
+  expect(diff2).toBeEmpty();
+  expect(diff1).bnEqual(sequencerRewards1);
+  expect(diff3).bnEqual(sequencerRewards3);
+  expect(sequencerRewards1).bnEqual(
+    sequencerRewards3.divn(2) || sequencerRewards3.divn(2).addn(1),
+  );
```
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:269]: When we divide the reward into three parts it may turn out that the reward is equal, then when dividing into 3 parts and 2 users we will get numbers that differ by one. I am not completely sure how this works in the system, but I received a reward for one and another sequencer, say, 630001 and 315000, while I wanted to leave the most correct check that these are not random numbers, namely 2/3 and 1/3 of the total reward

File: test/sequential/sequencerStaking.rewards.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (AlexChetverov) [reply test/sequential/sequencerStaking.rewards.test.ts:285]: Done

File: test/sequential/sequencerStaking.rewards.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (goncer) [reply test/sequential/sequencerStaking.rewards.test.ts:192]: ok! hopefully this goes right! :crossed_fingers:

**Review by goncer — APPROVED** [review]
